common/src/main/java/com/microsoft/hydralab/performance/inspectors/AndroidMemoryHprofInspector.java

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.hydralab.performance.inspectors;
+
+import com.microsoft.hydralab.common.util.ShellUtils;
+import com.microsoft.hydralab.common.util.TimeUtils;
+import com.microsoft.hydralab.performance.PerformanceInspection;
+import com.microsoft.hydralab.performance.PerformanceInspectionResult;
+import com.microsoft.hydralab.performance.PerformanceInspector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.Assert;
+
+import java.io.File;
+
+public class AndroidMemoryHprofInspector implements PerformanceInspector  {
+
+    private static final String RAW_RESULT_FILE_NAME_FORMAT = "%s_%s_%s_memory.hprof";
+    private static final String HPROF_FILE_PREFIX = "/data/local/tmp/";
+    private final Logger classLogger = LoggerFactory.getLogger(getClass());
+
+
+    @Override
+    public PerformanceInspectionResult inspect(PerformanceInspection performanceInspection) {
+
+        File rawResultFolder = new File(performanceInspection.resultFolder, performanceInspection.appId);
+        Assert.isTrue(rawResultFolder.exists() || rawResultFolder.mkdir(), "rawResultFolder.mkdirs() failed in" + rawResultFolder.getAbsolutePath());
+        String tmpTime = TimeUtils.getTimestampForFilename();
+        String hprofFileName = String.format(RAW_RESULT_FILE_NAME_FORMAT, getClass().getSimpleName(), performanceInspection.appId, tmpTime);
+        File rawResultFile = new File(rawResultFolder,
+                hprofFileName);
+        String sdHprofFilePath = HPROF_FILE_PREFIX + hprofFileName;
+        String dumpCommand = String.format(getMemHprofCommand(), performanceInspection.deviceIdentifier, performanceInspection.appId, sdHprofFilePath);
+
+        Process process = ShellUtils.execLocalCommand(dumpCommand, classLogger);
+        if (process != null && process.exitValue() == 0) {
+            Process pullProcess = ShellUtils.execLocalCommand("adb pull " + sdHprofFilePath + " " + rawResultFile.getAbsolutePath(), classLogger);
+            if (pullProcess != null && pullProcess.exitValue() == 0) {
+                return new PerformanceInspectionResult(rawResultFile, performanceInspection);
+            }
+        }
+        return new PerformanceInspectionResult(null, performanceInspection);
+    }
+
+    private String getMemHprofCommand() {
+        return "adb -s %s shell am dumpheap %s %s";
```

Comment:
How can we know whether the test app is debuggable build? Can we add a check here or in advance so that we should disable the heap dump for signed app.

Reply:
dumpheap has no restriction on debuggable or non-debuggable AFAIK, usually we don't dump heap in release build because dumpheap makes the JVM stuck, which affects the performance.

Reply:
Sounds good, I'll check the debuggable flag in this inspector.